### PR TITLE
Bluetooth: Mesh: Scheduler model implementation

### PR DESCRIPTION
**Pull request by alxelax**
Implements the following models:

Scheduler client
Scheduler server
Scheduler setup server

### CODE REVIEW

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+	sched_time = bt_mesh_time_srv_localtime(srv,
+			bt_mesh_time_srv_mktime(srv, sched_time));
+	if (item->day_of_week & (1 << sched_time->tm_wday)) {
+		return true;
+	}
+
+	if (item->day == ANY_DAY) {
+		int rest_wday = item->day_of_week >> sched_time->tm_wday;
+		int delta = rest_wday ? __builtin_ctz(rest_wday) :
+			__builtin_ctz(item->day_of_week) +
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:196]: Use `u32_count_trailing_zeros()` from zephyr/include/sys/math_extras.h

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+		return;
+	}
+
+	if (!set_minute(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	if (!set_second(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:328]: I will have to trust you on this algorithm I think. It'll need some pretty good test coverage.

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+		if (model_id == BT_MESH_MODEL_ID_GEN_ONOFF_SRV  &&
+				handled_mod != NULL) {
+			struct bt_mesh_onoff_srv *srv =
+			(struct bt_mesh_onoff_srv *)handled_mod->user_data;
+			struct bt_mesh_onoff_set set = {
+				.on_off =
+			metadata->srv->data.sch_reg[metadata->index].action,
+				.transition = &transition
+			};
+			srv->handlers->set(srv, NULL, &set, NULL);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:394]: This will bypass the publish mechanism in the onoff server. I think it's better to create an internal function like `bt_mesh_scene_srv_set` for onoff, and use that to call the set-handler.

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+		}
+
+		elem_index++;
+		elem = bt_mesh_elem_find(bt_mesh_primary_addr() + elem_index);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:398]:
```suggestion
		elem = bt_mesh_elem_find(elem->addr + 1);
```

Also, we need to verify that this is a unicast address before passing it to `find()`, in case the current element has address `0x3fff`

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 5]
```diff
@@ -0,0 +1,630 @@
...
+	if (item->second < current_local->tm_sec) {
+		return false;
+	}
+
+	bool ovflw = false;
+
+	if (item->second == EVERY_15_UNITS) {
+		sched_time->tm_sec = current_local->tm_sec < 15 ? 15 :
+				current_local->tm_sec < 30 ? 30 :
+				current_local->tm_sec < 45 ? 45 : 0;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:273]: `sched_time->tm_sec = 15 * ceiling_fraction(current_local->tm_sec, 15)`

File: tests/subsys/bluetooth/mesh/models/src/main.c [hunk 6]
```diff
@@ -98,7 +101,9 @@ static struct bt_mesh_elem elems[] = {
 		BT_MESH_MODEL_LIGHT_HSL_SRV(&light_hsl_srv),
 		BT_MESH_MODEL_LIGHT_HUE_SRV(&light_hue_srv),
 		BT_MESH_MODEL_LIGHT_SAT_SRV(&light_sat_srv),
-	), BT_MESH_MODEL_NONE),
+		BT_MESH_MODEL_SCENE_CLI(&scene_cli),
```
Comment (trond-snekvik) [inline tests/subsys/bluetooth/mesh/models/src/main.c:104]: There's already a scene client here.

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 7]
```diff
@@ -0,0 +1,630 @@
...
+	int days[12] = {31, is_leap_year(current_local->tm_year) ? 29 : 28,
+			31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+	struct tm revised_time = *current_local;
+
+	revised_time.tm_mday += number_days;
+
+	if (revised_time.tm_mday > days[revised_time.tm_mon]) {
+		revised_time.tm_mday -= days[revised_time.tm_mon];
+		revised_time.tm_mon = revised_time.tm_mon == DECEMBER ?
+				JANUARY : revised_time.tm_mon + 1;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:88]: In this case, revised_time needs to update its year as well. Does it?

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 8]
```diff
@@ -0,0 +1,630 @@
...
+			CONTAINER_OF(work,
+				struct bt_mesh_scheduler_scheduled_meta,
+				delayed_work);
+	struct bt_mesh_scheduler_scheduled_meta *metadata =
+			CONTAINER_OF(delayed_work,
+				struct bt_mesh_scheduler_scheduled_meta,
+				delayed_work);
+	metadata->srv->data.status_bitmap ^= 1u << metadata->index;
+
+	if (metadata->srv->data.sch_reg[metadata->index].action > 2) {
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/scheduler_srv.c:350]: Arguably, we didn't even have to schedule this. Same for invalid scene number

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: include/bluetooth/mesh/scheduler.h [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	/** Scheduled second for the action. */
+	uint16_t second : 6;
+	/** Schedule days of the week for the action. */
+	uint16_t day_of_week : 7;
+	/** Action to be performed at the scheduled time. */
+	uint16_t action : 4;
+	/** Transition time for this action. */
+	uint16_t transition_time : 8;
+	/** Scene number to be used for some actions. */
+	uint16_t scene_number : 16;
```
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler.h:46]: Unit in uint16_t gives the less sized structure. Removing the bitfield will not give any advantages. I'd suggest to leave as is for consistency.

File: include/bluetooth/mesh/scheduler.h [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+/** Scheduler status */
+struct bt_mesh_scheduler_status {
+	/** Bit field indicating defined Actions in the Schedule Register. */
+	uint16_t schedules;
+};
+
+/** Scheduler Action status */
+struct bt_mesh_scheduler_action_status {
+	/** Enumerates (selects) a Schedule Register entry. */
+	uint8_t index : 4;
```
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler.h:58]: Using bitfield allows to avoid value overflowing. Max index value is 16. It is not possible to set the wrong value if you have 4 bits wide type.

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 11]
```diff
@@ -0,0 +1,630 @@
...
+		 */
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_SCHEDULER_SETUP_SRV));
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_SCENE_SRV));
+		bt_mesh_model_extend(mod, bt_mesh_model_find(
+				bt_mesh_model_elem(mod),
+				BT_MESH_MODEL_ID_GEN_POWER_ONOFF_SRV));
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:596]: 5.3.6.1 Description
The Scheduler Setup Server model is used to support the configuration functionality of a node that
can store a schedule.
The Scheduler Setup Server is a main model that extends and corresponds with the Scheduler Server
model (see Section 5.3.5), and extends the Scene Setup Server (see Section 5.3.4) and the Generic
Power OnOff Setup Server (see Section 3.3.5) models.

File: include/bluetooth/mesh/scheduler.h [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+	uint16_t month : 12;
+	uint16_t day : 5;
+	uint16_t hour : 5;
+	uint16_t minute : 6;
+	uint16_t second : 6;
+	uint16_t day_of_week : 7;
+	uint16_t action : 4;
+	uint16_t transition_time : 8;
+	uint16_t scene_number : 16;
+};
```
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler.h:84]: I'm not sure I got this message. Does `net_buf_simple` provide endianness agnostic API for arbitrary sized data?

[the remark this answers is not in the document]
Comment (trond-snekvik) [reply include/bluetooth/mesh/scheduler.h:84]: It does not provide an API for bitfield data, no.

File: include/bluetooth/mesh/scheduler.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (trond-snekvik) [reply include/bluetooth/mesh/scheduler.h:58]: This overflows without warning for non-literal assignments though. If I make a function like 
```c
void schedule_action(uint8_t idx, const struct bt_mesh_scheduler_register *reg) 
{
    struct bt_mesh_scheduler_action_params params = {
        .idx = idx,
        .schedule_register = *reg,
    };
    bt_mesh_scheduler_cli_action_set(&sch_cli, NULL, &params, NULL);
}
```
it won't give me any warnings if I call `schedule_action(19, &weekdays);`, it'll just roll it over to send idx=3

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/scheduler_srv.c:596]: You're right. I posted a discussion in our team chat, I think this introduces a couple of issues.
The problem mentioned in the extension comment that's copied in all the setup server models also needs to be considered; we don't support models being extended by multiple other models at the moment. This can be addressed in upstream by switching to a non-hierarchical extension tree, like a circular linked list or something. We'll need to find a way to preserve the hierarchy later though.

File: include/bluetooth/mesh/scheduler_cli.h [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+ *  @retval -EALREADY A blocking request is already in progress.
+ *  @retval -EADDRNOTAVAIL A message context was not provided and publishing is
+ *  not configured.
+ *  @retval -EAGAIN The device has not been provisioned.
+ *  @retval -ETIMEDOUT The request timed out without a response.
+ */
+int bt_mesh_scheduler_cli_action_get(struct bt_mesh_scheduler_cli *cli,
+				struct bt_mesh_msg_ctx *ctx,
+				uint8_t index,
+				struct bt_mesh_scheduler_action_status *rsp);
```
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler_cli.h:119]: It is parsed data from the incoming message. If everything works well they are the same but it will skip malfunction behavior if a server replies to the wrong index.

File: subsys/bluetooth/mesh/scheduler_cli.c [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+	struct bt_mesh_scheduler_cli *cli = mod->user_data;
+	struct bt_mesh_scheduler_action_status action_status;
+
+	if (buf->len != BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS &&
+	    buf->len !=	BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS_REDUCED) {
+		return;
+	}
+
+	memset(&action_status, 0,
+			sizeof(struct bt_mesh_scheduler_action_status));
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_cli.c:47]: Not always. As I mentioned before there is possible to get reduced status. This is bad practice to rise garbage to app.

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 15]
```diff
@@ -0,0 +1,630 @@
...
+	srv->data.status_bitmap = 0;
+	memset(&srv->data.sch_reg, 0, sizeof(srv->data.sch_reg));
+
+	struct bt_mesh_model *time_mod =
+			bt_mesh_model_find(bt_mesh_model_elem(mod),
+			BT_MESH_MODEL_ID_TIME_SRV);
+	if (time_mod == NULL) {
+		return -ECANCELED;
+	}
+	srv->time_srv = (struct bt_mesh_time_srv *)time_mod->user_data;
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:576]: btw. it is not caught by `zephyr/scripts/checkpatch.pl` that checks diff according to Linux coding style.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 4 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:398]: I think, my case is better since I get the next element unequivocally. It does not require address checking since the primary address and element index related to each other linearly. I'd prefer to leave as is.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 5 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:273]: Nop. It gives a different result to what I have here. Just quickly checked in the online compiler. Values from 0 to 60.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 7 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:88]: you're right, it does.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 2 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:328]: I fully agree with you. This should be covered by unit tests as much as possible. There are mant corner cases in time conversion and rescheduling.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 3 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:394]: Looks like Scene model will not publish the status as well. I need to add here publishing.

File: include/bluetooth/mesh/scheduler.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler.h]: sys_byteorder swapping has been used depending on endianness. The scheduler model should be compatible with both little and big endian platforms.

File: include/bluetooth/mesh/scheduler.h [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+#include <bluetooth/mesh/model_types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define BT_MESH_MODEL_SCHEDULER_ACTION_ENTRY_NUMBER     16u
+
+/** Schedule Register. */
+struct bt_mesh_schedule_register {
```
[the remark this answers is not in the document]
Comment (alxelax) [reply include/bluetooth/mesh/scheduler.h]: Changed name on `struct bt_mesh_schedule_register` since register consists of entries according to spec

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 17]
```diff
@@ -0,0 +1,630 @@
...
+	if (!set_second(&sched_time, current_local, item, srv->time_srv)) {
+		return;
+	}
+
+	int64_t scheduled_uptime = bt_mesh_time_srv_mktime(srv->time_srv,
+			&sched_time);
+	k_delayed_work_submit(&metadata->delayed_work,
+		K_MSEC(scheduled_uptime > current_uptime ?
+				scheduled_uptime - current_uptime : 0));
+	srv->data.status_bitmap ^= 1u << index;
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c]: used `WRITE_BIT` instead

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 18]
```diff
@@ -0,0 +1,630 @@
...
+
+	revised_time.tm_min += number_minutes;
+	return set_minute(sched_time, &revised_time, item, srv);
+}
+
+static bool set_year(struct tm *sched_time,
+		     struct tm *current_local,
+		     struct bt_mesh_schedule_register *item)
+{
+	if (item->year < current_local->tm_year && item->year != ANY_YEAR) {
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c]: redesigned year setting. It is tolerant to any year rollover. However, if you provide an entry year less than the current year for instance on 1 year you will have to wait for 99 years :)

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/scheduler_srv.c]: I know :( It's a convention in the Bluetooth subsystem in upstream, that we've been trying to follow in downstream as well.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/scheduler_srv.c]: c'est la vie :man_shrugging:

File: include/bluetooth/mesh/scheduler.h [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+#define BT_MESH_SCHEDULER_OP_ACTION_STATUS    BT_MESH_MODEL_OP_1(0x5F)
+#define BT_MESH_SCHEDULER_OP_ACTION_SET       BT_MESH_MODEL_OP_1(0x60)
+#define BT_MESH_SCHEDULER_OP_ACTION_SET_UNACK BT_MESH_MODEL_OP_1(0x61)
+
+#define BT_MESH_SCHEDULER_MSG_LEN_ACTION_GET 1
+#define BT_MESH_SCHEDULER_MSG_LEN_ACTION_SET 10
+#define BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS 10
+#define BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS_REDUCED 1
+#define BT_MESH_SCHEDULER_MSG_LEN_GET 0
+#define BT_MESH_SCHEDULER_MSG_LEN_STATUS 2
```
Comment (trond-snekvik) [inline include/bluetooth/mesh/scheduler.h:136]: Wrap these defines in 
`/** @cond INTERNAL_HIDDEN */`
`/** @endcond */`
like this: https://github.com/nrfconnect/sdk-nrf/blob/master/include/bluetooth/mesh/time.h#L89-L113

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 4 above)
Comment (trond-snekvik) [reply subsys/bluetooth/mesh/scheduler_srv.c:403]: Elem find will still return a valid elem if you give it a group address, so I think both methods have the same problem

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
I have marked all resolved items accordingly. Some additional comments below.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 4 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:403]: Group address never can be the primary one. My case excludes this mistake :)

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/scheduler_srv.c:403]: That is true, but `bt_mesh_primary_addr() + elem_index` can be

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 20]
```diff
@@ -0,0 +1,749 @@
...
+		return;
+	}
+
+	srv->sch_reg[idx] = tmp;
+	BT_DBG("Rx: scheduler server action index %d set, ack %d", idx, ack);
+
+	if (srv->sch_reg[idx].action < BT_MESH_SCHEDULER_SCENE_RECALL ||
+	   (srv->sch_reg[idx].action == BT_MESH_SCHEDULER_SCENE_RECALL &&
+	    srv->sch_reg[idx].scene_number != 0)) {
+		schedule_action(srv, idx);
```
Comment (ludvigsj) [inline subsys/bluetooth/mesh/scheduler_srv.c:583]: I see you check this here, but even if `action > BT_MESH_SCHEDULER_SCENE_RECALL` the `srv->sch_reg[idx]` will be overwritten. Is this on purpose?

File: subsys/bluetooth/mesh/scheduler_cli.c [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+
+const struct bt_mesh_model_op _bt_mesh_scheduler_cli_op[] = {
+	{
+		BT_MESH_SCHEDULER_OP_STATUS,
+		BT_MESH_SCHEDULER_MSG_LEN_STATUS,
+		status_op,
+	},
+	{
+		BT_MESH_SCHEDULER_OP_ACTION_STATUS,
+		BT_MESH_SCHEDULER_MSG_LEN_ACTION_STATUS,
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_cli.c:78]: You're right. Thanks, good catch.

File: tests/subsys/bluetooth/mesh/models/src/main.c [hunk 22]
```diff
@@ -98,7 +101,8 @@ static struct bt_mesh_elem elems[] = {
 		BT_MESH_MODEL_LIGHT_HSL_SRV(&light_hsl_srv),
 		BT_MESH_MODEL_LIGHT_HUE_SRV(&light_hue_srv),
 		BT_MESH_MODEL_LIGHT_SAT_SRV(&light_sat_srv),
-	), BT_MESH_MODEL_NONE),
+		BT_MESH_MODEL_SCHEDULER_CLI(&scheduler_cli)
+		), BT_MESH_MODEL_NONE),
```
[the remark this answers is not in the document]
Comment (alxelax) [reply tests/subsys/bluetooth/mesh/models/src/main.c:105]: there was some merge conflict, will fix

File: subsys/bluetooth/mesh/scheduler_srv.c [hunk 23]
```diff
@@ -0,0 +1,749 @@
...
+	revised_time.tm_mday += number_days;
+
+	if (revised_time.tm_mday > days[revised_time.tm_mon]) {
+		revised_time.tm_mday -= days[revised_time.tm_mon];
+		if (revised_time.tm_mon == DECEMBER) {
+			revised_time.tm_mon = JANUARY;
+			revised_time.tm_year++;
+			is_year_revised = true;
+		} else {
+			revised_time.tm_mon++;
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:84]: This is not a universal function. It revises months for the specific case. It is when we have `any day` setting for `day of the month` but restrictions for `day of week`. If there is `any day` for `day of month` the model should choose the closest day from the allowed `day of week`.  The difference between the current day and the allowed day is calculated here:
https://github.com/nrfconnect/sdk-nrf/pull/3753/files#diff-147a66791a463d17316ec75175ef3be31199aa0e3c3cbb680b15aa8b6d8b17b2R223
It cannot be technically more than 6 days (week size - 1). It is not possible to overflow months twice.
The worst case is when there are no available days in the rest of the month. This is exactly covered by this function.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 20 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:583]: According to the standard client can set the entry configuration with the action `BT_MESH_SCHEDULER_NO_ACTIONS`  (0xf) and the Scheduler Server should return this entry configuration when client sends `Scheduler Action Get`.
However, there is no sense to schedule the register entry with action `no action`.

I'll add more strict action value checking

**Comment by alxelax** [comment]
Changed license header in the new files as well.

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 20 above)
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:585]: According to the standard client can set configuration entry with `no action` and the server shall store it. The client can get this configuration back. Added more strict action checking.

File: subsys/bluetooth/mesh/scheduler_cli.c (hunk 21 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_cli.c]: good catch, thnx

File: subsys/bluetooth/mesh/scheduler_srv.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/scheduler_srv.c:84]: This is not universal function. It is used only for case when we have setting `any day` for `day of month` but restrictions for `day of week`. Server should choose the closest available `day of week`. Number of days for revising cannot be more than `week size - 1` (6). It is not possible to overflow months twice.

**Review by ludvigsj — APPROVED** [review]

**Review by trond-snekvik — APPROVED** [review]

**Comment by trond-snekvik** [comment]
Ignoring compliance check failure for complex macro. This is by design.
